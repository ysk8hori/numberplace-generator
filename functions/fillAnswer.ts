import { pipe, tap } from 'remeda';
import {
  type Answer,
  type Cell,
  fillCellAnswer,
  filterByGroup,
  findCell,
  isCellsPosition,
  removeAnswerCandidate,
} from '../models/cell.ts';
import type { Position } from '../models/position.ts';
import { isNot } from 'remeda';

export const fillAnswer: (
  cells: Cell[],
) => (p: Position) => (a: Answer) => void = (cl) => (p) => (a) =>
  pipe(
    findCell(cl)(p),
    // 答えを記入
    tap(fillCellAnswer(a)),
    // ターゲットと同じグループのセルから候補を除去
    tap((c) =>
      c.groups
        .map(filterByGroup(cl.filter(isNot(isCellsPosition(p)))))
        .map(removeAnswerCandidate)
        .forEach((f) => f(a))
    ),
  );
