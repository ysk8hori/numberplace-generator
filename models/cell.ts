import {
  difference,
  isDeepEqual,
  isEmpty,
  isNonNullish,
  isNullish,
  map,
  pipe,
  piped,
  tap,
  unique,
  uniqueWith,
  zip,
} from 'remeda';
import type { Group } from './group.ts';
import { createPositions, isSamePos, type Position } from './position.ts';
import { type BlockSize, createGameRange } from './blockSize.ts';
import {
  getBlockGroup,
  getHorizontalGroup,
  getVerticalGroup,
} from './group.ts';
import { branch, throwError } from '../utils/utils.ts';
import { filter } from 'remeda';
import { isStrictEqual } from 'remeda';

export type Answer = number;
export type AnswerCandidate = number;

/** Internal type */
export type Cell = {
  pos: Position;
  answerMut: undefined | Answer;
  answerCnadidatesMut: AnswerCandidate[];
  groups: Group[];
};

export const createCells: (blocksize: BlockSize) => Cell[] = (blocksize) =>
  pipe(
    blocksize,
    createGameRange,
    createPositions,
    map(
      (pos) => ({
        answerMut: undefined,
        answerCnadidatesMut: createGameRange(blocksize),
        groups: [
          getVerticalGroup(pos),
          getHorizontalGroup(pos),
          getBlockGroup(blocksize)(pos),
        ],
        pos,
      } satisfies Cell),
    ),
  );

export const isCellsPosition: (p: Position) => (c: Cell) => boolean =
  (p) => (c) => isSamePos(c.pos, p);

export const isInGroup: (cell: Cell) => (group: Group) => boolean =
  (c) => (g) => c.groups.includes(g);

export const filterByGroup: (cells: Cell[]) => (group: Group) => Cell[] =
  (cl) => (g) => cl.filter((c) => isInGroup(c)(g));

export const getEmptyCells: (cells: Cell[]) => Cell[] = (cl) =>
  cl.filter((c) => isNonNullish(c.answerMut));

export const getByPosition: (cells: Cell[]) => (p: Position) => Cell = (cl) =>
(
  p,
) => cl.find(isCellsPosition(p)) ?? throwError(`該当するセルがない. ${p}`);

/** 指定したセルから指定した候補値を削除する。削除した場合は true を返し、削除対象がなかった場合は false を返す。*/
const removeAnswerCandidateForCell: (
  c: Cell,
) => (a: AnswerCandidate) => boolean = (cellMut) => (a) => {
  const oldAnswerCandidateList = cellMut.answerCnadidatesMut;
  cellMut.answerCnadidatesMut = cellMut.answerCnadidatesMut.filter(
    (v) => v !== a,
  );
  return oldAnswerCandidateList.length !== cellMut.answerCnadidatesMut.length;
};

/**
指定したセル全てから指定した候補値を削除する。候補値を削除したセルのリストを返却する。
削除対象としたセルのリストを返却する理由は、本関数によって変更が生じたか否かを確かめるのに使用するため。
*/
export const removeAnswerCandidate: (
  c: Cell[],
) => (a: AnswerCandidate) => Cell[] = (cells) => (a) =>
  pipe(
    cells,
    map(removeAnswerCandidateForCell),
    map((f) => f(a)),
    zip(cells),
    filter(([b, _]) => b),
    map(([_, c]) => c),
  );

/**
セルに答えを入力する。
副作用として、答えの候補値のリストも入力した答えのみを持つ状態にする。
*/
export const fillCellAnswer: (a: Answer) => (cellMut: Cell) => void = (a) =>
  piped(
    tap((c) => (c.answerMut = a)),
    tap((c) => (c.answerCnadidatesMut = [a])),
  );

export const findCell: (cells: Cell[]) => (p: Position) => Cell = (cl) => (p) =>
  cl.find((c) => isSamePos(c.pos, p)) ?? throwError(`該当するセルがない. ${p}`);

export const 同じ答え候補をもつセルの数が答え候補の数と同じか: (
  /** 同一グループのセル */
  cellsByGroup: Cell[],
) => (al: AnswerCandidate[]) => boolean = (cl) => (al) =>
  al.length === cl.filter((c) => isDeepEqual(c.answerCnadidatesMut, al)).length;

/** 同一グループのセルが持つ候補のリストのユニークなリストを取得する */
export const getUniqueAnswercandidatesList: (
  /** 同一グループのセル */
  cellsByGroup: Cell[],
) => AnswerCandidate[][] = (cl) =>
  uniqueWith(
    cl.map((c) => c.answerCnadidatesMut),
    isDeepEqual,
  );

const getUniqueGroups: (cells: Cell[]) => Group[] = (cl) =>
  unique(cl.flatMap((c) => c.groups));

export const セルが回答済みか: (cell: Cell) => boolean = (c) =>
  isNonNullish(c.answerMut);
export const セルが未回答か: (cell: Cell) => boolean = (c) =>
  isNullish(c.answerMut);
export const 未回答のセルを抽出する: (cells: Cell[]) => Cell[] = (cl) =>
  cl.filter(セルが未回答か);
/**
指定されたセル全部に対し、回答できる状態であることを確認する。
答えが記入されていないのに候補が空であるセルは回答できない不正な状態になっているとみなせる。
*/
export const 全てのセルが回答可能か: (cells: Cell[]) => boolean = (cl) =>
  !cl
    .filter((c) => isNullish(c.answerMut))
    .some((c) => isEmpty(c.answerCnadidatesMut));

export const 全てのセルが回答済みか: (cells: Cell[]) => boolean = (cl) =>
  cl.filter((c) => isNullish(c.answerMut)).length === 0;

/**
したものを利用する場合は refineAnswerCandidateRecursive を使うべき。

候補値の最適化を行う。指定した複数のセルの候補値を、できる限り絞り込む。

指定した複数のセルの候補値についてそれぞれグループに分け、同一グループに属するセルに以下を行う。

- 同じ候補値のリストを持つセルの数がその候補値のリスト長と同じ場合は、それらを持つセル達のみにそれらの候補値が当てはまるはずであるため、他のセルからそれらの候補値を削除する。
*/
const refineAnswerCandidate: (cellsGroupMut: Cell[]) => void = (cl) =>
  pipe(
    cl,
    getUniqueGroups,
    map(filterByGroup(cl)),
    map((cl) =>
      getUniqueAnswercandidatesList(cl)
        .filter(同じ答え候補をもつセルの数が答え候補の数と同じか(cl))
        .forEach((al) =>
          cl
            .filter((c) => !isDeepEqual(al, c.answerCnadidatesMut))
            .forEach(
              (
                c,
              ) => (c.answerCnadidatesMut = difference(
                c.answerCnadidatesMut,
                al,
              )),
            )
        )
    ),
  );

/**
候補値の最適化を行う。指定した複数のセルの候補値を、できる限り絞り込む。

指定した複数のセルの候補値についてそれぞれグループに分け、同一グループに属するセルに以下を行う。

- 同じ候補値のリストを持つセルの数がその候補値のリスト長と同じ場合は、それらを持つセル達のみにそれらの候補値が当てはまるはずであるため、他のセルからそれらの候補値を削除する。
*/
export const refineAnswerCandidateRecursive: (cellsGroupMut: Cell[]) => void = (
  cl,
) =>
  branch(
      (cl: Cell[]) => JSON.stringify(cl.map((cl) => cl.answerCnadidatesMut)),
      piped(tap(refineAnswerCandidate), (cl) =>
        JSON.stringify(cl.map((cl) => cl.answerCnadidatesMut))),
      isStrictEqual,
    )(cl)
    ? undefined
    : refineAnswerCandidateRecursive(cl);

export function cellToString(cells: Cell[]): string {
  const horizontalLines: Map<number, Cell[]> = cells.reduce(
    (p, cell) => {
      const lineNo = cell.pos[1];
      if (!p.has(lineNo)) p.set(lineNo, new Array<Cell>());
      p.get(lineNo)?.push(cell);
      return p;
    },
    new Map<number, Cell[]>(),
  );
  return Array.from(horizontalLines.values())
    .map((line) => line.map((cell) => cell.answerMut ?? ' ').join(','))
    .join('\n');
}
