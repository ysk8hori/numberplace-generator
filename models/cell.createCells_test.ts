import { assertSnapshot } from "@std/testing/snapshot";
import { type Cell, createCells, 回答できないセルがあるか } from "./cell.ts";
import { assert, assertEquals, assertFalse } from "@std/assert";
import { fillAnswer } from "../functions/fillAnswer.ts";

Deno.test("9x9サイズ position", async (t) => {
  const cells = createCells({ height: 3, width: 3 });
  assertEquals(cells.length, 81);
  assertEquals<Cell>(cells[80], {
    pos: { x: 8, y: 8 },
    groups: ["v8", "h8"],
    answerMut: undefined,
    answerCnadidatesMut: [0, 1, 2, 3, 4, 5, 6, 7, 8],
  });
});

Deno.test("9x9サイズ snapshot", async (t) => {
  const cells = createCells({ height: 3, width: 3 });
  await assertSnapshot(t, cells);
});

Deno.test(
  "3x3サイズのセルを生成する",
  async (t) => await assertSnapshot(t, createCells({ height: 1, width: 3 })),
);

Deno.test("未回答セルの答え候補がなくなった場合は回答できない", () => {
  const cells = createCells({ height: 1, width: 3 });
  assertFalse(回答できないセルがあるか(cells));
  const fillAnswerByPos = fillAnswer(cells);

  fillAnswerByPos({ x: 1, y: 1 })(0);
  assertFalse(回答できないセルがあるか(cells));
  fillAnswerByPos({ x: 1, y: 1 })(1);
  assertFalse(回答できないセルがあるか(cells));
  fillAnswerByPos({ x: 1, y: 1 })(2);
  assert(回答できないセルがあるか(cells));
});
