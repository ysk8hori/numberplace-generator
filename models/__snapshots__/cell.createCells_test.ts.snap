export const snapshot = {};

snapshot[`9x9サイズ snapshot 1`] = `
[
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v0",
      "h0",
    ],
    pos: [
      0,
      0,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v1",
      "h0",
    ],
    pos: [
      1,
      0,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v2",
      "h0",
    ],
    pos: [
      2,
      0,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v3",
      "h0",
    ],
    pos: [
      3,
      0,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v4",
      "h0",
    ],
    pos: [
      4,
      0,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v5",
      "h0",
    ],
    pos: [
      5,
      0,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v6",
      "h0",
    ],
    pos: [
      6,
      0,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v7",
      "h0",
    ],
    pos: [
      7,
      0,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v8",
      "h0",
    ],
    pos: [
      8,
      0,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v0",
      "h1",
    ],
    pos: [
      0,
      1,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v1",
      "h1",
    ],
    pos: [
      1,
      1,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v2",
      "h1",
    ],
    pos: [
      2,
      1,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v3",
      "h1",
    ],
    pos: [
      3,
      1,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v4",
      "h1",
    ],
    pos: [
      4,
      1,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v5",
      "h1",
    ],
    pos: [
      5,
      1,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v6",
      "h1",
    ],
    pos: [
      6,
      1,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v7",
      "h1",
    ],
    pos: [
      7,
      1,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v8",
      "h1",
    ],
    pos: [
      8,
      1,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v0",
      "h2",
    ],
    pos: [
      0,
      2,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v1",
      "h2",
    ],
    pos: [
      1,
      2,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v2",
      "h2",
    ],
    pos: [
      2,
      2,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v3",
      "h2",
    ],
    pos: [
      3,
      2,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v4",
      "h2",
    ],
    pos: [
      4,
      2,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v5",
      "h2",
    ],
    pos: [
      5,
      2,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v6",
      "h2",
    ],
    pos: [
      6,
      2,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v7",
      "h2",
    ],
    pos: [
      7,
      2,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v8",
      "h2",
    ],
    pos: [
      8,
      2,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v0",
      "h3",
    ],
    pos: [
      0,
      3,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v1",
      "h3",
    ],
    pos: [
      1,
      3,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v2",
      "h3",
    ],
    pos: [
      2,
      3,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v3",
      "h3",
    ],
    pos: [
      3,
      3,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v4",
      "h3",
    ],
    pos: [
      4,
      3,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v5",
      "h3",
    ],
    pos: [
      5,
      3,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v6",
      "h3",
    ],
    pos: [
      6,
      3,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v7",
      "h3",
    ],
    pos: [
      7,
      3,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v8",
      "h3",
    ],
    pos: [
      8,
      3,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v0",
      "h4",
    ],
    pos: [
      0,
      4,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v1",
      "h4",
    ],
    pos: [
      1,
      4,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v2",
      "h4",
    ],
    pos: [
      2,
      4,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v3",
      "h4",
    ],
    pos: [
      3,
      4,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v4",
      "h4",
    ],
    pos: [
      4,
      4,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v5",
      "h4",
    ],
    pos: [
      5,
      4,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v6",
      "h4",
    ],
    pos: [
      6,
      4,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v7",
      "h4",
    ],
    pos: [
      7,
      4,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v8",
      "h4",
    ],
    pos: [
      8,
      4,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v0",
      "h5",
    ],
    pos: [
      0,
      5,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v1",
      "h5",
    ],
    pos: [
      1,
      5,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v2",
      "h5",
    ],
    pos: [
      2,
      5,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v3",
      "h5",
    ],
    pos: [
      3,
      5,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v4",
      "h5",
    ],
    pos: [
      4,
      5,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v5",
      "h5",
    ],
    pos: [
      5,
      5,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v6",
      "h5",
    ],
    pos: [
      6,
      5,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v7",
      "h5",
    ],
    pos: [
      7,
      5,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v8",
      "h5",
    ],
    pos: [
      8,
      5,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v0",
      "h6",
    ],
    pos: [
      0,
      6,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v1",
      "h6",
    ],
    pos: [
      1,
      6,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v2",
      "h6",
    ],
    pos: [
      2,
      6,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v3",
      "h6",
    ],
    pos: [
      3,
      6,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v4",
      "h6",
    ],
    pos: [
      4,
      6,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v5",
      "h6",
    ],
    pos: [
      5,
      6,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v6",
      "h6",
    ],
    pos: [
      6,
      6,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v7",
      "h6",
    ],
    pos: [
      7,
      6,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v8",
      "h6",
    ],
    pos: [
      8,
      6,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v0",
      "h7",
    ],
    pos: [
      0,
      7,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v1",
      "h7",
    ],
    pos: [
      1,
      7,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v2",
      "h7",
    ],
    pos: [
      2,
      7,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v3",
      "h7",
    ],
    pos: [
      3,
      7,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v4",
      "h7",
    ],
    pos: [
      4,
      7,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v5",
      "h7",
    ],
    pos: [
      5,
      7,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v6",
      "h7",
    ],
    pos: [
      6,
      7,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v7",
      "h7",
    ],
    pos: [
      7,
      7,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v8",
      "h7",
    ],
    pos: [
      8,
      7,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v0",
      "h8",
    ],
    pos: [
      0,
      8,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v1",
      "h8",
    ],
    pos: [
      1,
      8,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v2",
      "h8",
    ],
    pos: [
      2,
      8,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v3",
      "h8",
    ],
    pos: [
      3,
      8,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v4",
      "h8",
    ],
    pos: [
      4,
      8,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v5",
      "h8",
    ],
    pos: [
      5,
      8,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v6",
      "h8",
    ],
    pos: [
      6,
      8,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v7",
      "h8",
    ],
    pos: [
      7,
      8,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
    ],
    answerMut: undefined,
    groups: [
      "v8",
      "h8",
    ],
    pos: [
      8,
      8,
    ],
  },
]
`;

snapshot[`3x3サイズのセルを生成する 1`] = `
[
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
    ],
    answerMut: undefined,
    groups: [
      "v0",
      "h0",
    ],
    pos: [
      0,
      0,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
    ],
    answerMut: undefined,
    groups: [
      "v1",
      "h0",
    ],
    pos: [
      1,
      0,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
    ],
    answerMut: undefined,
    groups: [
      "v2",
      "h0",
    ],
    pos: [
      2,
      0,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
    ],
    answerMut: undefined,
    groups: [
      "v0",
      "h1",
    ],
    pos: [
      0,
      1,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
    ],
    answerMut: undefined,
    groups: [
      "v1",
      "h1",
    ],
    pos: [
      1,
      1,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
    ],
    answerMut: undefined,
    groups: [
      "v2",
      "h1",
    ],
    pos: [
      2,
      1,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
    ],
    answerMut: undefined,
    groups: [
      "v0",
      "h2",
    ],
    pos: [
      0,
      2,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
    ],
    answerMut: undefined,
    groups: [
      "v1",
      "h2",
    ],
    pos: [
      1,
      2,
    ],
  },
  {
    answerCnadidatesMut: [
      0,
      1,
      2,
    ],
    answerMut: undefined,
    groups: [
      "v2",
      "h2",
    ],
    pos: [
      2,
      2,
    ],
  },
]
`;
